### DISCUSSION TITLE
Keycloak only for roles

### DISCUSSION THREAD

**Discussion by gagostin** [keycloak/keycloak#14674] [Ideas]
It looks like I got stuck and I need some help on solving this authentication/authorization problem.
I have an Okta installation, outside of my control, which supports OIDC but doesn't allow to easily manage roles and also does not support the token exchange flow.
I was hoping to use Keycloak as sort of sidecar/federated system to store users' roles, but I still need to perform the authentication against Okta, simply because that's the one storing the user credentials and I cannot have direct access to those.
Also, I cannot expose Keycloak to the end users, due to security constraints.
In other words, the client can only reach Okta and must authenticate on Okta, but since I can't store user roles on Okta I wish to use Keycloak for that and query Keycloak from my backend application to query for user roles, all of that without using the token exchange flow....

Any ideas how can I achieve that, if possible at all?

**Comment by pckhoi** [comment]
That sounds like [identity provider integration](https://www.keycloak.org/docs/latest/server_admin/#_identity_broker) to me. If what Keycloak provides is not enough, looks like you can [extend its identity brokering functionality](https://www.keycloak.org/docs/latest/server_development/#identity-brokering-apis) too. I have not dabbled much in this area so I can't offer anything more specific.
